byron/ledger/executable-spec/test/Ledger/Relation/Properties.hs

```diff
@@ -0,0 +1,223 @@
+  -> Set (Domain r)
+  -> r
+  -> r
+  -> m ()
+propDomainRestrictionAndUnion find x s r1 r2
+  = find x (s ◁ (r1 ∪ r2))
+    ===
+    find x ((s ◁ r1) ∪ (s ◁ r2))
+-}
+
+--------------------------------------------------------------------------------
+-- Functions abstracted over 'Relation r'
+--------------------------------------------------------------------------------
+
+-- | Combined property abstracted over 'Relation r'
+propRelation_
+  :: ( Relation r, Ord (Domain r), Ord (Range r)
+     , Show r, Show (Domain r), Show (Range r))
+  => Gen r
+  -> Gen [Domain r]
+  -> (Domain r -> r -> Maybe (Range r))
+  -> PropertyT IO ()
+propRelation_ genRelation_ genList_ find_
+  = do
+      r1 <- forAll $ genRelation_
+      s <- forAll $ Set.fromList <$> genList_
+      -- pick an element from 's ∪ dom r'
+      x <- forAll $ Gen.element (Set.toList $ s `Set.union` dom r1)
+
+      classify "(dom r) ∩ s == dom (s ◁ r)" $
+               Set.member x (dom r1 `Set.intersection` s)
+      propDomainRestrictionAndIntersection x s r1
+
+      classify "(dom r ∩ s) ◁ r == s ◁ r" $
+               Set.member x (dom ((dom r1 `Set.intersection` s) ◁ r1))
+      propDomainRestrictionAndIntersectionB find_ x s r1
+
+      classify "(dom r \\ s) ◁ r == s ⋪ r" $
+               Set.member x (dom ((dom r1 `Set.difference` s) ◁ r1))
+      propDomainExclusionAndSetDifference find_ x s r1
+
+      r2 <- forAll $ genRelation_
+      -- pick an element from 's ∪ dom r1 ∪ dom r2'
+      x' <- forAll $ Gen.element (Set.toList $ s `Set.union` dom r1 `Set.union` dom r2)
+
+      classify "(dom r1 ∪ s) ⋪ (r1 ∪ r2) == (dom r1 ∪ s) ⋪ r2" $
+               Set.member x' (dom ((dom r1 `Set.union` s) ⋪ (r1 ∪ r2)))
+      propDomainExclusionAndUnion find_ x' s r1 r2
+
+      {- TODO @uroboros fails for Bimap, but could make it pass with
+         more a more restrictive Bimap generator.
+
+      classify "s ◁ (r1 ∪ r2) == (s ◁ r1) ∪ (s ◁ r2)" $
+                Set.member x' (dom (s ◁ (r1 ∪ r2)))
+      propDomainRestrictionAndUnion find_ x' s r1 r2
+      -}
+
+genList :: Gen [Int]
+genList = Gen.list (Range.constant 5 100)
+                   (Gen.integral (Range.constant 0 50))
+
+genTuples :: Gen [(Int, Int)]
+genTuples = zip <$> genList <*> genList
+
+--------------------------------------------------------------------------------
+-- Test 'instance Relation Map'
+--------------------------------------------------------------------------------
+
+genMap :: Gen (Map Int Int)
+genMap = Map.fromList <$> genTuples
+
+propRelationMap :: Property
+propRelationMap
+  = withTests 500 $ property $ propRelation_ genMap genList find
+    where
+      find :: Int -> Map Int Int -> Maybe Int
+      find = Map.lookup
+
+--------------------------------------------------------------------------------
+-- Test 'instance Relation Set'
+--------------------------------------------------------------------------------
+
+genSet :: Gen (Set (Int, Int))
+genSet = Set.fromList <$> genTuples
```

Comment:
Why not using [`set`](http://hackage.haskell.org/package/hedgehog-1.0/docs/Hedgehog-Gen.html#v:set)